### PR TITLE
Welcome emails

### LINKED ISSUE #354: Welcome emails
I'm finally at the point where I'd like to automate a welcome email to new developers and businesses to the site. To start, I'd love to keep these as unstyled HTML and very short.

A PR addressing this issue will:

- [ ] Add an `after_create_commit` callback to both `Developer` and `Business` that sends an email
    * Subject: Welcome to railsdevs!
    * Body: (see below)
    * From: Joe Masilotti joe@masilotti.com (formatted correctly)
- [ ] Ensure the HTML and plain text versions work when previewing
- [ ] Add a test for each that ensures the email is enqueued correctly
- [ ] **BONUS**: Implement [`name_of_person`](https://github.com/basecamp/name_of_person) to use the developer/business's first name

---

### Developer welcome email

Hey #{developer.name},

I'm Joe, the founder of [railsdevs](https://railsdevs.com). I saw that you added your developer profile and wanted to reach out.

What does your ideal next role look like?

### Business welcome email

Hey #{business.name},

I'm Joe, the founder of [railsdevs](https://railsdevs.com). I saw that you added your business profile and wanted to reach out. If there's anything I can do to make your search easier let me know.

What are you looking for in a Rails developer?

### PR DESCRIPTION
**Pull request by tim-carey-code**
Closes #354 
Send a welcome email when a Business or Developer is created. Added a callback in the Business and Developer models to send an email after creation. Made a new mailer(WelcomeMailer) to handle the emails. Added HTML and text previews for both emails. Added tests in business_tests and developer_tests to ensure emails are enqueued. 

### Pull request checklist

- [x] Your code contains tests relevant for code you modified
- [x] You have linted and tested the project with `bin/check`
- [x] You have added significant changes and product updates to the [changelog](CHANGELOG.md)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 ### April
 
+* April 15 - Automatically send out welcome emails on Business and Developer creation
```
Comment (joemasilotti) [inline CHANGELOG.md:7]:
```suggestion
* April 15 - Automatically send out welcome emails on Business and Developer creation - #378
```

File: app/mailers/welcome_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class WelcomeMailer < ApplicationMailer
+  default from: "joe@masilotti.com"
+  helper :messages
```
Comment (joemasilotti) [inline app/mailers/welcome_mailer.rb:3]: Both of these can be removed as they are already in `ApplicationMailer`.

File: app/mailers/welcome_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class WelcomeMailer < ApplicationMailer
```
Comment (joemasilotti) [inline app/mailers/welcome_mailer.rb:1]: I'd rather see these live in `DeveloperMailer` and `BusinessMailer`. Can you please refactor to accommodate that?

File: app/views/welcome_mailer/business_welcome_email.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<body>
+  <p>Hey <%= @business.name %>,</p>
+  <p>
+    I'm Joe the founder of <%= link_to "railsdevs", root_url %>. I saw that you
+    added your business profile and wanted to reach out. If there's anything I
+    can do to make your search easier let me know.
```
Comment (joemasilotti) [inline app/views/welcome_mailer/business_welcome_email.html.erb:6]: Please move this to a single line.

File: app/views/welcome_mailer/business_welcome_email.text.erb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Hey <%= @business.name %>,
+
+I'm Joe the founder of railsdevs. I saw that you
+added your business profile and wanted to reach out. If there's anything I
+can do to make your search easier let me know.
```
Comment (joemasilotti) [inline app/views/welcome_mailer/business_welcome_email.text.erb:5]: Please move this to a single line.

File: app/views/welcome_mailer/developer_welcome_email.html.erb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<p>Hey <%= @developer.name %>,</p>
+<p>
+  I'm Joe the founder of <%= link_to "railsdevs", root_url %>. I saw that you
+  added your developer profile and wanted to reach out.
+</p>
```
Comment (joemasilotti) [inline app/views/welcome_mailer/developer_welcome_email.html.erb:5]: Please move this to a single line.

File: app/views/welcome_mailer/developer_welcome_email.text.erb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+Hey <%= @developer.name %>,
+
+I'm Joe the founder of railsdevs. I saw that you
+added your developer profile and wanted to reach out.
```
Comment (joemasilotti) [inline app/views/welcome_mailer/developer_welcome_email.text.erb:4]: Please move this to a single line.

File: app/views/welcome_mailer/developer_welcome_email.text.erb [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+
+What does your ideal next role look like?
```
Comment (joemasilotti) [inline app/views/welcome_mailer/developer_welcome_email.text.erb:6]:
```suggestion
What does your ideal next role look like?

https://railsdevs.com
```

File: test/fixtures/welcome_mailer/welcome_business_text [hunk 9]
```diff
@@ -0,0 +1,7 @@
+Hey Business Owner with Conversation,
+
+I'm Joe the founder of railsdevs. I saw that you
+added your business profile and wanted to reach out. If there's anything I
+can do to make your search easier let me know.
+
+What are you looking for in a Rails developer?
```
Comment (joemasilotti) [inline test/fixtures/welcome_mailer/welcome_business_text:7]: This file can be deleted.

File: test/fixtures/welcome_mailer/welcome_developer_text [hunk 10]
```diff
@@ -0,0 +1,6 @@
+Hey Prospect Developer,
+
+I'm Joe the founder of railsdevs. I saw that you
+added your developer profile and wanted to reach out.
+
+What does your ideal next role look like?
```
Comment (joemasilotti) [inline test/fixtures/welcome_mailer/welcome_developer_text:6]: This file can be deleted.

File: test/mailers/previews/welcome_mailer_preview.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# Preview all emails at http://localhost:3000/rails/mailers/welcome_mailer
```
Comment (joemasilotti) [inline test/mailers/previews/welcome_mailer_preview.rb:1]: Please remove this comment.

File: test/mailers/welcome_mailer_test.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
+require "test_helper"
+
+class WelcomeMailerTest < ActionMailer::TestCase
```
Comment (joemasilotti) [inline test/mailers/welcome_mailer_test.rb:3]: Please remove the comments in this file.

File: test/mailers/welcome_mailer_test.rb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    # Create the email and store it for further assertions
+    email = WelcomeMailer.with(developer: @developer).developer_welcome_email
+
+    # Send the email, then test that it got queued
+    assert_emails 1 do
+      email.deliver_now
+    end
+
+    # Test the body of the sent email contains what we expect it to
+    assert_equal ["joe@masilotti.com"], email.from
```
Comment (joemasilotti) [inline test/mailers/welcome_mailer_test.rb:15]: Please remove this assertion.

File: test/mailers/welcome_mailer_test.rb [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    assert_equal [@developer.user.email], email.to
+    assert_equal "Welcome to railsdevs!", email.subject
```
Comment (joemasilotti) [inline test/mailers/welcome_mailer_test.rb:17]: Please remove this assertion.

File: test/mailers/welcome_mailer_test.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    assert_equal read_fixture("welcome_developer_text").join, email.text_part.body.to_s.strip
```
Comment (joemasilotti) [inline test/mailers/welcome_mailer_test.rb:18]: Please remove this assertion in favor of checking for a few words inside of the email body.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking good! I left a few inline comments and suggestions. Outside of that, can you please: remove the changes to `Brewfile.lock.json`?

Note that I asked about refactoring the naming of `WelcomeEmail` to `BusinessEmail` and `DeveloperEmail` - that needs to be applied across the codebase.

Also, the changes I suggested in the tests are to make them less brittle. I don't want them to fail if we change a bit of copy (or translate them in the future).

Let me know if that doesn't makes or if you have questions!

File: app/mailers/business_mailer.rb [hunk 16]
```diff
@@ -8,4 +8,9 @@ def developer_profiles
 
     mail(to: @business.user.email, subject:)
   end
+
+  def business_welcome_email
```
Comment (joemasilotti) [inline app/mailers/business_mailer.rb:12]:
```suggestion
  def welcome_email
```

File: app/mailers/developer_mailer.rb [hunk 17]
```diff
@@ -9,4 +9,9 @@ def invisiblize
 
     mail(to: recipient.email, subject: "Your profile on railsdevs was flagged")
   end
+
+  def developer_welcome_email
```
Comment (joemasilotti) [inline app/mailers/developer_mailer.rb:13]:
```suggestion
  def welcome_email
```

File: app/models/business.rb [hunk 18]
```diff
@@ -13,11 +13,15 @@ class Business < ApplicationRecord
...
-  after_create_commit :send_admin_notification
+  after_create_commit :send_admin_notification, :send_business_welcome_email
 
   private
 
   def send_admin_notification
     NewBusinessNotification.with(business: self).deliver_later(User.admin)
   end
+
+  def send_business_welcome_email
```
Comment (joemasilotti) [inline app/models/business.rb:24]:
```suggestion
  def send_welcome_email
```

File: app/models/developer.rb [hunk 19]
```diff
@@ -91,6 +91,10 @@ def send_admin_notification
     NewDeveloperProfileNotification.with(developer: self).deliver_later(User.admin)
   end
 
+  def send_developer_welcome_email
```
Comment (joemasilotti) [inline app/models/developer.rb:94]:
```suggestion
  def send_welcome_email
```

File: test/mailers/developer_mailer_test.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  test "welcome developer email test" do
+    @developer = Developer.first
+    email = DeveloperMailer.with(developer: @developer).developer_welcome_email
+
+    assert_emails 1 do
+      email.deliver_now
+    end
+
+    assert_equal [@developer.user.email], email.to
+    assert_not_nil email.body.to_s
```
Comment (joemasilotti) [inline test/mailers/developer_mailer_test.rb:13]:
```suggestion
    assert_match /added your developer profile/, email.body.to_s
```

File: test/mailers/business_mailer_test.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+  test "welcome business email test" do
+    @business = Business.first
+    email = BusinessMailer.with(business: @business).business_welcome_email
+
+    assert_emails 1 do
+      email.deliver_now
+    end
+
+    assert_equal [@business.user.email], email.to
+    assert_not_nil email.body.to_s
```
Comment (joemasilotti) [inline test/mailers/business_mailer_test.rb:13]:
```suggestion
    assert_match /added your business profile/, email.body.to_s
```

**Review by joemasilotti — CHANGES_REQUESTED** [review]
Can you fix the line endings of the marked files? They have a little red circle with a line in the GitHub diff.

Outside of that and some nitpicky code style changes this looks good to merge!

File: test/mailers/developer_mailer_test.rb (hunk 20 above)
Comment (tim-carey-code) [reply test/mailers/developer_mailer_test.rb:13]: Test won't pass unless I use email.body.encoded instead, is this OK?

Comment (joemasilotti) [reply test/mailers/developer_mailer_test.rb:13]: Yeah go for it!

**Comment by joemasilotti** [comment]
Thanks so much for the contribution! I'll merge this in as soon as CI passes from my copy-edit commit.

**Comment by tim-carey-code** [comment]
I'm glad I was able to contribute! It was a pleasure working with you! Hope to contribute more in the future

**Pull request merged**